README
                               Set::Relation
                                  README
---------------------------------------------------------------------------

by Darren Duncan <darren@DarrenDuncan.net>

PREFACE

This distribution features the Perl module named Set::Relation (a relation
data type) and its support structure.

Read below for how to get support or keep up to date.

DOCUMENTATION READING ORDER

If you want to read all the Set::Relation documentation that exists now in
this distribution, I recommend doing so in this order:

   1.       README
   2. (opt) Changes
   3.       lib/Set/Relation.pm
   4.       lib/Set/Relation/V1.pm
   5.       lib/Set/Relation/V2.pm
   6. (opt) TODO

Then you can carry on with anything else.

You may prefer the nicely HTML-formatted versions of the above, which are
available at "http://search.cpan.org/dist/Set-Relation/".

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my personal website, "http://www.darrenduncan.net/",
or through my business website, "https://muldis.com".

For those of you that are interested in the bleeding edge of development, I
have all parts of the Set::Relation project in one or more public
repositories.  The Perl Set::Relation project is in a Git repository
at "https://github.com/muldis/Set_Relation_Perl".

You can download slightly older versions of all of my modules from the
public CPAN archives at "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
on my own web server at "http://www.darrenduncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is at
"http://search.cpan.org", and my own DUNCAND author page is at
"http://search.cpan.org/search?author=DUNCAND".

The main CPAN search site url for Set::Relation for Perl is
"http://search.cpan.org/dist/Set-Relation/"; that is the url you should
link to presently if you want something to consider the "home page" of
Set::Relation; it is best to refer people there by default.

You can download my oldest tarred and gzipped distributions from BACKPAN,
at "http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my
web server, at "http://www.darrenduncan.net/d/perl/archives/", right back
to the libdwg 1.0, first released on 2000 July 23.

FIN

Share and Enjoy!


Changes
                               Set::Relation
                                  Changes
---------------------------------------------------------------------------

20xx-xx-xx

    * Set::Relation version 0.13.4 for Perl is released on CPAN as
    Set-Relation-0.13.4.tar.gz.

    * Eliminate the test-only external dependency on the CPAN module
    Carp::Always as it is not actually needed for testing at all.
    As a result, running Set::Relation's tests no longer requires having
    these 5 non-core direct or non-direct dependencies to be installed:
    Carp::Always, Test::Base, Spiffy, Algorithm::Diff, Text::Diff.
    After this change, Test::Deep is the only test-only external dependency
    remaining that isn't a Perl core module.

2023-09-25

    * Set::Relation version 0.13.3 for Perl is released on CPAN as
    Set-Relation-0.13.3.tar.gz.

    * The main purpose of this release is to make major updates to
    Set::Relation's documentation.  A lot of archaic and superfluous and
    obsolete documentation was purged.  A small amount of outdated
    documentation was corrected.  There were absolutely zero code changes
    made, so the actual Set::Relation functionality is fully unchanged.

    * The 2 files "INSTALL" and "TODO" were entirely removed.  The vast
    majority of the file "README" was removed; some "README" fixes were made.

    * Remove all references to a Raku version of Set::Relation developed in
    parallel with the Perl version; we only talk about the Perl version here.

    * Remove all references to official Mailman mailing lists for support
    and discussion as these no longer exist.

    * Update the main module description text to better reflect the current
    and intended maintenance plans for the project.

    * Make the distro pass the Perl Kwalitee "has_abstract_in_pod" test:
    For every pod file, tweak its NAME section so the abstract
    is on the same line as the package name rather than the next line.

    * Update main module name/abstract text to "A relation data type"
    from "Relation data type for Perl".

    * Update all references to GitHub to use the updated project url
    "https://github.com/muldis/Set_Relation_Perl".

    * Replaced any references to the non-existent CPAN distribution/package
    name Muldis::D::RefEng with the title "Muldis Data Engine Reference"
    so that these don't constitute broken external links.
    Replaced any references to the non-existent CPAN distribution/package
    name Muldis::D with the title "Muldis Data Language" so that these
    don't constitute broken external links.  Likewise replace any
    references to the "Muldis D" language with the same.

    * Incremented all copyright year range-ends to 2023.

2019-10-20

    * Set::Relation version 0.13.2 for Perl is released on CPAN as
    Set-Relation-0.13.2.tar.gz.

    * Split the 3 files lib/Set/Relation[|/V[1|2]].pm into themselves plus
    the 3 new files lib/Set/Relation[|/V[1|2]].pod or in other words split
    every main file in 2.  As such, all of the lib/*.pm files now only
    contain Perl code, and all POD now lives in lib/*.pod files.

    * Between the prior release and this one, the programming language
    formerly known as "Perl 6" was renamed to "Raku".  As such, all
    documentation that refers to that language was updated to use its new
    name, and all language that referred to "Perl 5" now simply refers to
    "Perl" because "Perl" now unambiguously means Perl 5.  The only
    exception to these changes is in documentation about history, such as
    all prior entries in this Changes file.

    * (Relation.pod, V2.pm, V1.pm)  Minor docs update to reflect that the
    obsolete Muldis D dialect HDMD_Perl5_STD was renamed to HDMD_Perl_STD.

    * Lowered the declared minimum Perl version dependency to 5.8.0 (2002)
    from 5.8.1 (2003) since the prior .1 requirement seemed rather
    arbitrary and all 5.8.z and newer should suffice.

    * The declared minimum version dependencies for some Perl-bundled
    modules were lowered to those shipped with 5.8.0 from those with 5.8.1:
    ExtUtils::MakeMaker is now 6.03 from 6.17.

    * For the first time, as long as your Perl is version 5.26.0 (2017) or
    newer, each of the 2 Set::Relation classes requires only the Perl core
    and core modules to run, and nothing else, so usage is trivially easy.
    If your Perl is an older version, back to 5.8.1 (2003), you need only
    also install the 1 dual-life CPAN module List::Util version 1.4602+.
    If your Perl is instead version 5.8.0 (2002), you need to install that
    plus the 1 dual-life CPAN module Scalar::Util version 1.13+.

    * Increased the explicit dependency on the Perl core module List::Util
    to 1.4602 (was 1.45), which is the version bundled with Perl 5.26.0;
    1.45 is the actual minimum dependency, but 1.4602 was chosen so it
    matched the version with the first stable Perl release to have any
    suitable version.  Similarly for Scalar::Util, 1.09 is the actual
    minimum dependency, but 1.13 matches what was bundled with Perl 5.8.1.

    * Reversing a minor change by release 0.13.0, increased the explicit
    dependency on the Perl core module Test::More to 0.92 (was 0.47), which
    is the version bundled with Perl 5.10.1; the reason is that the test
    suite uses the done_testing() feature added by Test::More version 0.88.

    * Set::Relation is now released under the terms of the Artistic
    License version 2 rather than the GNU Lesser General Public License
    (LGPL) version 3+.

    * Updated all documentation to use an updated canonical Muldis Data
    Systems url of https://muldis.com rather than several http:// variants.

    * Incremented all copyright year range-ends to 2019.

2016-06-26

    * Set::Relation version 0.13.1 for Perl 5 is released on CPAN as
    Set-Relation-0.13.1.tar.gz.

    * The main purpose of this release is to excise Set::Relation's last

Makefile.PL
use 5.008000;
use utf8;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# Make setting optional MakeMaker parameters more readable.
sub OPTIONAL {
   return () unless $ExtUtils::MakeMaker::VERSION ge shift;
   return @_;
}

WriteMakefile(
   NAME => 'Set::Relation',
   VERSION => '0.13.3',
   ABSTRACT => 'A relation data type',
   AUTHOR => 'Darren Duncan <darren@DarrenDuncan.net>',
   PREREQ_PM => {
      # This list is a copy of "META_ADD:[configure_|build_|]requires"
      # where the other is canonical; this list is for use by MakeMaker
      # itself, which in general wouldn't see the info in the other lists,
      # as the other lists are primarily used when generating the META.yml/json.
      'Carp' => '1.01',
      'ExtUtils::MakeMaker' => '6.03',
      'List::Util' => '1.4602',
      'Scalar::Util' => '1.13',
      'Test::Deep' => '0.106',
      'Test::More' => '0.92',
   },
   OPTIONAL( '6.48',
      MIN_PERL_VERSION => '5.008000',
   ),
   OPTIONAL( '6.31',
      LICENSE => 'artistic_2',
   ),
   OPTIONAL( '6.46',
      # Use META_ADD instead of META_MERGE so that we can remove
      # any build-time dependencies that MakeMaker will put into
      # the requires field.  (Comment From DBD::SQLite; do *we* need that?)
      META_ADD => {
         configure_requires => {
            'ExtUtils::MakeMaker' => '6.03', # bundled with Perl 5.8.0
         },
         build_requires => {
            'ExtUtils::MakeMaker' => '6.03', # bundled with Perl 5.8.0
            # These are bundled with newer versions of Perl only.
            'Test::More' => '0.92', # bundled with Perl 5.10.1, min 0.88 needed
            # These are not bundled with any version of Perl.
            'Test::Deep' => '0.106', # tests req 0.106+ w Perl 5.10.1+
         },
         requires => {
            'perl' => '5.008000', # as MIN_PERL_VERSION no affect META.yml
            'Carp' => '1.01', # bundled with Perl 5.8.0
            # These are bundled with newer versions of Perl only.
            'Scalar::Util' => '1.13', # bundled with Perl 5.8.1, min 1.09 needed
            'List::Util' => '1.4602', # bundled with Perl 5.26.0, min 1.45 needed
         },
         resources => {
            homepage => 'https://muldis.com',
            license => 'https://www.perlfoundation.org/artistic-license-20.html',
            repository => 'https://github.com/muldis/Set_Relation_Perl',
         },
      },
   ),
);

1;
